# Indel caller performance in commercial and clinical dilution series
Cedric Chauve, March 3rd, 2021.

## Data

This report describes the results of the new indel caller on a series of
commercial and clinical dilution series.

Commercial samples:
- `CG001Qv51Next002, 201002_NB551381_0087_AHFW2FAFX2, "3 controls, 4ng and 8ng and 16ng dilution series of horizon cfDNA at 1%, 0.7%, 0.5%, 0.25%, 0.1%, 0%"`
- `CG001Qv51Next007, 201023_NB551381_0093_AHFWLJAFX2, "3 controls, 4ng and 8ng and 16ng dilution series of horizon cfDNA at 1%, 0.7%, 0.5%, 0.25%, 0.1%, 0%""`
- `CG001Qv51Next024, 210121_NB551381_0110_AHJM5HAFX2, "SeraSeq ctDNA reference standards, 4ng and 8ng and 16ng dilution series of horizon cfDNA at 1%, 0.75%, 0.5%, 0.375%, 0.25%, 0.125%, 0%"`
- `CG001Qv51Next025, 210125_NB551381_0111_AHJLY3AFX2, "SeraSeq ctDNA reference standards, 4ng and 8ng and 16ng dilution series of horizon cfDNA at 1%, 0.75%, 0.5%, 0.375%, 0.25%, 0.125%, 0%"`
- `CG001Qv51Next026, 210129_NB551381_0114_AHKFMWAFX2, "SeraSeq ctDNA reference standards, 4ng and 8ng and 16ng dilution series of horizon cfDNA at 1%, 0.75%, 0.5%, 0.375%, 0.25%, 0.125%, 0%"`

Clinical samples
- `CG001Qv51Next012,201104_NB551381_0098_AHHKK5AFX2, "3 controls, 3 dilution series at 4ng, 8ng, 12ng"`
- `CG001Qv51Next032,210205_NB551381_0118_AHMHT3AFX2, "3 controls, 7 VAFs (28%, 5%, 2.5%, 1%, 0.75%, 0.5%, 0%) at 4ng, 7 VAFs at 8ng, 7 VAFs at 12ng"`
- `CG001Qv51Next034,210217_NB551381_0120_AHMHG7AFX2, "4ng, 8ng, 16ng at 5%, 3.5%, 2.5%, 1%, 0.75%, 0.5%, 0%"`

Prior to the analysis, a list of 6 indels was identified in these samples that
appear in all samples. These indels were *blacklisted* in further analysis, and
their properties in commercial and clinical samples are described below.

Commercial samples (ctrl is the ratio between the indel VAF in a sample and in
the normal female):
```
indel                   nb_occ  min_vaf max_vaf mean_vaf  min_ctrl  max_ctrl  mean_ctrl
chr3:178936116:GT:C     90      44.79   50.85   47.8      0.83      0.95      0.89
chr19:4117563:GCTT:G    90      1.07    3.03    1.89      0.03      0.15      0.09
chr15:90631917:TC:T     90      0.64    31.59   10.29     0.03      1.0       0.58
chr12:25398297:C:CCG    90      3.18    9.44    5.41      0.41      0.9       0.61
chr12:25380252:TC:T     90      0.07    8.1     2.25      0.1       1.0       0.66
chr1:156846307:GC:G     90      0.1     26.06   9.66      0.0       1.0       0.49
```

Clinical samples:
```
indel                   nb_occ  min_vaf max_vaf mean_vaf  min_ctrl  max_ctrl  mean_ctrl
chr3:178936116:GT:C     48      33.47   46.81   42.75     0.9       1.0       0.97
chr19:4117563:GCTT:G    48      0.54    1.82    1.06      0.08      0.27      0.13
chr15:90631917:TC:T     48      0.66    1.3     0.95      0.68      1.0       0.95
chr12:25398297:C:CCG    48      0.91    4.94    3.54      0.41      1.0       0.74
chr12:25380252:TC:T     48      0.06    0.87    0.18      0.58      1.0       0.96
chr1:156846307:GC:G     48      0.1     0.2     0.14      0.55      1.0       0.9
```

An important feature of these blacklisted indels is that they generally appear
also in the normal female.


## Methods

Each indel detected by the indel caller was scored as follows:
    **0.5 x complexity + support + overlap + control**
where each term is a penalty between 0 and 1 based on the following features:
- complexity: sequence complexity around the breakpoint,
- support: ratio of supporting reads in the largest supporting cluster,
- overlap: ratio of the cumulated VAF of overlapping indels and  the indel VAF,
- control: ratio of the VAF of the indel in the normal female and of the indel
  VAF.  
The maximum score is 3.5.

A detected indel was called if
- its weighted score is at most 0.9,
- the overlap and control penalties are both < 1.0.
The parameters 0.5 and 0.9 were chosen after an extensive parameters grid
exploration.

## Results

The table below shows the general statistics for.

### Commercial runs: wildtype samples

The indel caller does not call any of the expected indels in the wildtype
samples.

### Commercial runs: non-wildtype samples

The two tables presents the results based first on the Horizon runs of
commercial samples, then on the SeraSeq runs. `FN_u` is for "unobersved FN",
i.e. an expected indel that does not appear in any alignment, while `FN_o` is
for "observed FN", i.e. an expected indel that is observed but not called.

```
# Horizon samples
vaf_values      ng_range        score   w_comp  TP      FP      TN      FN      FN_u    FN_o    sens.   spec.   acc.    youden  prec.   recall  F1      FDR
0.1_0.125       4.0_64.0        0.9     0.5     6       1146    11887   6       6       0       0.5     0.912   0.912   0.412   0.005   0.5     0.01    0.995
0.25            4.0_64.0        0.9     0.5     6       50      1739    6       3       3       0.5     0.972   0.969   0.472   0.107   0.5     0.176   0.893
0.5             4.0_64.0        0.9     0.5     20      2       485     4       1       3       0.833   0.996   0.988   0.829   0.909   0.833   0.869   0.091
0.7_0.75        4.0_64.0        0.9     0.5     10      0       103     2       0       2       0.833   1.0     0.983   0.833   1.0     0.833   0.909   0.0
1.0             4.0_64.0        0.9     0.5     11      0       64      1       0       1       0.917   1.0     0.987   0.917   1.0     0.917   0.957   0.0
#
0.1_0.125       8.0_64.0        0.9     0.5     5       784     7757    3       3       0       0.625   0.908   0.908   0.533   0.006   0.625   0.012   0.994
0.25            8.0_64.0        0.9     0.5     5       23      1023    3       1       2       0.625   0.978   0.975   0.603   0.179   0.625   0.278   0.821
0.5             8.0_64.0        0.9     0.5     14      0       250     2       0       2       0.875   1.0     0.992   0.875   1.0     0.875   0.933   0.0
0.7_0.75        8.0_64.0        0.9     0.5     7       0       56      1       0       1       0.875   1.0     0.984   0.875   1.0     0.875   0.933   0.0
1.0             8.0_64.0        0.9     0.5     7       0       32      1       0       1       0.875   1.0     0.975   0.875   1.0     0.875   0.933   0.0
```

```
# SeraSeq samples
vaf_values      ng_range        score   w_comp  TP      FP      TN      FN      FN_u    FN_o    sens.   spec.   acc.    youden  prec.   recall  F1      FDR
0.1_0.125       4.0_64.0        0.9     0.5     31      77      2021    23      19      4       0.574   0.963   0.954   0.537   0.287   0.574   0.383   0.713
0.25            4.0_64.0        0.9     0.5     31      112     2631    23      16      7       0.574   0.959   0.952   0.533   0.217   0.574   0.315   0.783
0.375           4.0_64.0        0.9     0.5     37      10      428     17      11      6       0.685   0.977   0.945   0.662   0.787   0.685   0.732   0.213
0.5             4.0_64.0        0.9     0.5     40      8       305     14      10      4       0.741   0.974   0.94    0.715   0.833   0.741   0.784   0.167
0.7_0.75        4.0_64.0        0.9     0.5     41      10      159     13      9       4       0.759   0.941   0.897   0.7     0.804   0.759   0.781   0.196
1.0             4.0_64.0        0.9     0.5     43      10      80      11      9       2       0.796   0.889   0.854   0.685   0.811   0.796   0.803   0.189
#
0.1_0.125       8.0_64.0        0.9     0.5     24      45      1168    12      10      2       0.667   0.963   0.954   0.63    0.348   0.667   0.457   0.652
0.25            8.0_64.0        0.9     0.5     23      62      1484    13      8       5       0.639   0.96    0.953   0.599   0.271   0.639   0.381   0.729
0.375           8.0_64.0        0.9     0.5     26      3       244     10      7       3       0.722   0.988   0.954   0.71    0.897   0.722   0.8     0.103
0.5             8.0_64.0        0.9     0.5     28      5       177     8       6       2       0.778   0.973   0.94    0.751   0.848   0.778   0.811   0.152
0.7_0.75        8.0_64.0        0.9     0.5     28      6       84      8       6       2       0.778   0.933   0.889   0.711   0.824   0.778   0.8     0.176
1.0             8.0_64.0        0.9     0.5     29      7       46      7       6       1       0.806   0.868   0.843   0.674   0.806   0.806   0.806   0.194
```

The new indel caller makes close to no error in the Horizon runs, aside of a
few FN. The FN are two indels:
- `chr7:55242464:AGGAATTAAGAGAAGC:A` is not observed in the sequencing data in
  samples with low expected VAF,
- `chr7:55248998:A:ATGGCCAGCG` is not observed in a few low expected VAF samples.  
Both, when observed are not called because they occur in amplicons with
multiple co-optimal alignments which creates a high overlap penalty.

The two FP called in Horizon samples at expected VAF >= 0.5 are
`chr10:43609002:AG:A` and `chr4:55593653:AACACAAC:A`.

The FP called in SeraSeq samples at expected VAF >= 0.375% are more numerous
but are dominated by an indel `chr10:89717769:TA:T` that represents 47 out of
59 FP. The other FP are
- `chr1:156849061:GGCCACACGCAACTGTCTAGTGGGCCAGGGACTGGTGGTCAAGATTGGTGATTTTGGCATGAGCAGGGATATCTA:G`
  (3 occurrences, in SeraSeq run CG001Qv51Next026, VAF in 3%-6%),
- `chr11:534294:CCCA:C` (2 occurrences, same aliquot, low VAF),
- `chr4:55595540:AT:A` (2 occurrences, low VAF),
- `chrX:66941757:AC:A` (2 occurrences, low VAF),
- `chr17:7579513 TG:T` (2 occurrences, low VAF),
- `chr7:137029834:CTT:C` (1 occurrence).

### Commercial runs: QMRS samples

The indel called in the QMRS include the expected indel
 `chr7:55242464:AGGAATTAAGAGAAGC:A`(called in all QMRS
samples) plus some other indels, some of which are in the blacklisted indels:
- `chr12:25380252:TC:T` (VAF 10%, blacklisted indel)
- `chr19:4117563:GCTT:G` (VAF ~1%, blacklisted indel)
- `chr3:41266133:CCTT:C` (VAF 10%)
- `chr1:156846307:GC:G` (VAF > 30%),
- `chrX:27130922:TA:T` (VAF > 10%),

### Clinical runs: wildtype samples

Here too, no expected indel is called and FP are low VAF indels.

### Clinical runs: non-wildtype samples

```
vaf_values      ng_range        score   w_comp  TP      FP      TN      FN      FN_u    FN_o    sens.   spec.   acc.    youden  prec.   recall  F1      FDR
0.0_0.25        4.0_12.0        0.9     0.5     13      238     5330    11      5       6       0.542   0.957   0.955   0.499   0.052   0.542   0.095   0.948
0.25_0.5        4.0_12.0        0.9     0.5     16      24      1612    0       0       0       1.0     0.985   0.985   0.985   0.4     1.0     0.571   0.6
0.5_1.0         4.0_12.0        0.9     0.5     24      0       791     0       0       0       1.0     1.0     1.0     1.0     1.0     1.0     1.0     0.0
1.0_5.0         4.0_12.0        0.9     0.5     50      3       389     0       0       0       1.0     0.992   0.993   0.992   0.943   1.0     0.971   0.057
5.0_100.0       4.0_12.0        0.9     0.5     18      7       24      0       0       0       1.0     0.774   0.857   0.774   0.72    1.0     0.837   0.28
#
0.0_0.25        8.0_12.0        0.9     0.5     10      129     3399    9       4       5       0.526   0.963   0.961   0.489   0.072   0.526   0.127   0.928
0.25_0.5        8.0_12.0        0.9     0.5     15      15      1043    0       0       0       1.0     0.986   0.986   0.986   0.5     1.0     0.667   0.5
0.5_1.0         8.0_12.0        0.9     0.5     23      0       443     0       0       0       1.0     1.0     1.0     1.0     1.0     1.0     1.0     0.0
1.0_5.0         8.0_12.0        0.9     0.5     47      3       241     0       0       0       1.0     0.988   0.99    0.988   0.94    1.0     0.969   0.06
5.0_100.0       8.0_12.0        0.9     0.5     16      6       10      0       0       0       1.0     0.625   0.812   0.625   0.727   1.0     0.842   0.273
#
0.0_0.25        12.0_12.0       0.9     0.5     5       55      1634    4       2       2       0.556   0.967   0.965   0.523   0.083   0.556   0.144   0.917
0.25_0.5        12.0_12.0       0.9     0.5     10      10      479     0       0       0       1.0     0.98    0.98    0.98    0.5     1.0     0.667   0.5
0.5_1.0         12.0_12.0       0.9     0.5     11      0       171     0       0       0       1.0     1.0     1.0     1.0     1.0     1.0     1.0     0.0
1.0_5.0         12.0_12.0       0.9     0.5     22      3       118     0       0       0       1.0     0.975   0.979   0.975   0.88    1.0     0.936   0.12
5.0_100.0       12.0_12.0       0.9     0.5     8       2       6       0       0       0       1.0     0.75    0.875   0.75    0.8     1.0     0.889   0.2
```

The precision (PPV) for the VAF range 5%-100% is strongly impacted by the low
number of data points.

The FN that are observed but not called are `chr7:55248998:A:ATGGCCAGCG` and
`chr17:7578457:CG:C` are not called due to a combination of low VAF (<0.5%) and
high overlap penalty.

The FP are
- `chrX:27130922:TA:T` (15 out of 73 FP), also observed in the QMRS samples,
- `chr10:43613840:G:GC` (34 occurrences, low VAF, also observed many times at
  low VAF in the commercial samples),
- `chr1:156849061:GGCCACACGCAACTGTCTAGTGGGCCAGGGACTGGTGGTCAAGATTGGTGATTTTGGCATGAGCAGGGATATCTA:G` (9 occurrences),
- `chr7:137029834:CTT:C` (4 occurrences, low VAF),
- `chr10:43609017:TG:T` (3 occurrences, low VAF),
- `chr7:55228049:G:GC` (3 occurrences, low VAF),
- plus a handful of FP occurring just once.

### Clinical runs: QMRS samples

Here too, the expected indels are called, together with the same FPs than for
commercial samples.

## FFPE QMRS samples

The following runs were analyzed, each with three QMRS samples of various
qualities (mild, moderate and severe):
```
CG001Qv51Next008, 201026_NB551381_0094_AHHH73AFX2
CG001Qv51Next009, 201027_NB551381_0095_AHHGYHAFX2
```

In all 6 QMRS samples, the expected indel, as specified by the vendor, was
detected with appropriate VAF and the same other indels than in the plasma QMRS
samples were detected, plus a handful of other low VAF (<0.5%) indels.
